File: day02.rs
use common::get_input;
use regex::Regex;

type Input = Vec<((usize, usize), String, String)>;


fn part1(input: &Input) -> usize {
    input.iter()
        .filter(|((a, b), letter, password)| {
            let count = password.matches(letter).count();
            a <= &count && &count <= b
        })
        .count()
}

fn matches_once<T: PartialEq>(letter: T, a: T, b: T) -> bool {
    a != b && (letter == a || letter == b)
}

fn part2(input: &Input) -> usize {
    input.iter()
        .filter(|((a, b), letter, password)| {
            let chars = password.as_str();
            matches_once(letter.as_str(), &chars[(a-1)..*a], &chars[(b-1)..*b])
        })
        .count()
}

macro_rules! parse {
    ( $cap:expr, $i:expr, $type:ty ) => {
        $cap.get($i).unwrap().as_str().parse::<$type>().unwrap()
    }
}

fn parse(text: String) -> Input {
    let re = Regex::new(r"^(\d+)-(\d+) (.+): (.+)$").unwrap();
    text.trim()
        .split('\n')
        .map(|line| {
            let cap = re.captures(line).unwrap();
            ((parse!(cap, 1, usize), parse!(cap, 2, usize)), parse!(cap, 3, String), parse!(cap, 4, String))
        })
        .collect()
}

fn main() {
    let input = parse(get_input(02, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}

File: day02.py
"""Solution to day 2 of Advent of Code"""

from get_input import get_input, line_parser

import re


def exactly_one(value, a, b):
    return a != b and (a == value or b == value)


def part1(lines):
    return sum(1 for ((a, b), letter, password) in lines
               if a <= password.count(letter) <= b)


def part2(lines):
    return sum(1 for ((a, b), letter, password) in lines
               if exactly_one(letter, password[a-1], password[b-1]))


def parse(line):
    m = re.match(r'^(\d+)-(\d+) (.+): (.+)$', line)
    small, big, letter, password = m.groups()
    return ((int(small), int(big)), letter, password)


if __name__ == "__main__":
    LINES = line_parser(get_input(day=2, year=2020), parse=parse)
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")
